spyder/utils/icon_manager.py

```diff
@@ -458,9 +468,13 @@ def get_icon_by_extension_or_type(fname, scale_factor):
                     if bin_name in application_icons:
                         icon_by_extension = icon(
                             application_icons[bin_name], scale_factor)
+    icon_dict[(extension, scale_factor)] = icon_by_extension
     return icon_by_extension
 
 
+get_icon_by_extension_or_type._icon_dict = {}
```

Comment:
I don't like this because it's very pythonic, i.e. adding a hidden attribute to a function. Instead please declare a constant at the top of this file (e.g. `ICONS_BY_EXTENSION = {}`) and use it to cache and retrieve icons in `get_icon_by_extension_or_type`.